File: src/input_ffi.mjs
import fs from "node:fs";
import { Buffer } from "node:buffer"
import { Ok, Error as GError } from "./gleam.mjs";

/**
 * Mimics Python's `input` builtin.
 * 
 * @param {string} prompt
 * @returns {string} 
 */
export function input(prompt) {
  process.stdout.write(prompt);

  // 4096 bytes is the limit for cli input in bash.
  const buffer = Buffer.alloc(4096); 
  const bytesRead = fs.readSync(0, buffer, 0, buffer.length, null);
  const input = buffer.toString('utf-8', 0, bytesRead).trim();

  if (input) {
    return new Ok(input.trim());
  } else {
    return new GError(undefined);
  }
}

File: src/input_ffi.erl
-module(input_ffi).
-export([input/1]).

% Mimics Python's builtin `input` function.
-spec input(io:prompt()) -> {ok, unicode:unicode_binary()} | {error, eof | no_data}.
input(Prompt) ->
    case io:get_line(Prompt) of
        eof -> {error, eof};
        {error, _} -> {error, no_data};
        Data when is_binary(Data) -> {ok, string:trim(Data, trailing, "\r\n")};
        Data when is_list(Data) -> {ok, string:trim(unicode:characters_to_binary(Data), trailing, "\r\n")}
    end.